pkg/sdkserver/sdkserver.go

```diff
@@ -486,9 +469,53 @@ func (s *SDKServer) WatchGameServer(_ *sdk.Empty, stream sdk.SDK_WatchGameServer
 }
 
 // Reserve moves this GameServer to the Reserved state for the Duration specified
-// TODO: implement this functionality
-func (s *SDKServer) Reserve(_ context.Context, d *sdk.Duration) (*sdk.Empty, error) {
-	return &sdk.Empty{}, errors.New("not implemented")
+func (s *SDKServer) Reserve(ctx context.Context, d *sdk.Duration) (*sdk.Empty, error) {
+	s.stopReserveTimer()
+
+	e := &sdk.Empty{}
+
+	s.logger.Info("Received Reserve request, adding to queue")
+	s.enqueueState(stablev1alpha1.GameServerStateReserved)
+
+	// 0 is forever.
+	if d.Seconds == 0 {
+		return e, nil
+	}
+
+	duration := time.Duration(d.Seconds) * time.Second
+	s.gsUpdateMutex.Lock()
```

Comment:
should this be done before the enqueueState call above? It seems racey to set the new state and the duration in two separate locks regardless of the order....